conn.go

```diff
@@ -2,52 +2,127 @@ package conn
+// singleConn represents a single stream-multipexed connection to another Peer (IPFS Node).
 type singleConn struct {
-	local  peer.ID
-	remote peer.ID
-	maconn tpt.Conn
-	event  io.Closer
-}
+	streamConn smux.Conn
+	tptConn    tpt.Conn
+
+	secSession secio.Session
+
+	event io.Closer
+}
+
+var _ iconn.Conn = &singleConn{}
+
+// newSingleConn constructs a new connection
+func newSingleConn(ctx context.Context, local, remote peer.ID, privKey ci.PrivKey, tptConn tpt.Conn, pstpt smux.Transport, isServer bool) (iconn.Conn, error) {
+	ml := lgbl.Dial("conn", local, remote, tptConn.LocalMultiaddr(), tptConn.RemoteMultiaddr())
+
+	var streamConn smux.Conn
+	var secSession secio.Session
+	switch conn := tptConn.(type) {
+	case tpt.DuplexConn:
+		c := conn
+		// 1. secure the connection
+		if privKey != nil && iconn.EncryptConnections {
+			var err error
+			secSession, err = setupSecureSession(ctx, local, privKey, conn)
+			if err != nil {
+				return nil, err
+			}
+			c = &secureConn{
+				DuplexConn: conn,
+				secure:     secSession,
+			}
+		} else {
+			log.Warning("creating INSECURE connection %s at %s", tptConn.LocalMultiaddr(), tptConn.RemoteMultiaddr())
+		}
 
-// newConn constructs a new connection
-func newSingleConn(ctx context.Context, local, remote peer.ID, maconn tpt.Conn) (iconn.Conn, error) {
-	ml := lgbl.Dial("conn", local, remote, maconn.LocalMultiaddr(), maconn.RemoteMultiaddr())
+		// 2. start stream multipling
+		var err error
+		streamConn, err = pstpt.NewConn(c, isServer)
+		if err != nil {
+			return nil, err
+		}
+	case tpt.MultiplexConn:
+		// 1. secure the connection on the first best stream
+		if privKey != nil && iconn.EncryptConnections {
+			var stream smux.Stream
+			var err error
+			if isServer {
+				stream, err = conn.AcceptStream()
+			} else {
+				stream, err = conn.OpenStream()
+			}
+			if err != nil {
+				return nil, err
+			}
+			defer stream.Close()
+
+			secSession, err = setupSecureSession(ctx, local, privKey, stream)
+			if err != nil {
+				return nil, err
+			}
+		}
+		streamConn = conn
+	}
 
 	conn := &singleConn{
-		local:  local,
-		remote: remote,
-		maconn: maconn,
-		event:  log.EventBegin(ctx, "connLifetime", ml),
+		streamConn: streamConn,
+		tptConn:    tptConn,
+		secSession: secSession,
+		event:      log.EventBegin(ctx, "connLifetime", ml),
 	}
 
 	log.Debugf("newSingleConn %p: %v to %v", conn, local, remote)
 	return conn, nil
 }
 
+func setupSecureSession(ctx context.Context, local peer.ID, privKey ci.PrivKey, ch io.ReadWriteCloser) (secio.Session, error) {
+	if local == "" {
+		return nil, errors.New("local peer is nil")
+	}
+	if privKey == nil {
+		return nil, errors.New("private key is nil")
+	}
+	sessgen := secio.SessionGenerator{
+		LocalID:    local,
+		PrivateKey: privKey,
+	}
+	secSession, err := sessgen.NewSession(ctx, ch)
+	if err != nil {
+		return nil, err
+	}
+	// force the handshake right now
+	// TODO: find a better solution for this
+	b := []byte("handshake")
```

Comment:
I had to force the handshake to get QUIC working. I'm not entirely sure why it didn't work otherwise (I never understood why the secio handshake is delayed), I'd have to debug this again.

Since QUIC support will not be merged before TLS 1.3, we might not need this at all. I will just remove this code for now.

Reply:
It's a bit more complicated. The code won't work for duplex conns if we remove this here, because the context passed to secio might be cancelled before the handshake completes.

I'm wondering why we need this delayed handshake at all, it doesn't give us any benefit here, but makes it a lot harder to reason about timeouts (if the handshake is delayed, we can't have a connection establishment timeout).

Reply:
I'm not sure why we do this lazily but this definitely isn't the solution (it would, at the very least, require a protocol version bump to avoid breaking existing implementations).

I believe you can currently force a secio handshake by calling `secSession.ReadWriter()` (ideally, you'd be able to write an empty byte string to force the issue but we tend not to handle those very well in libp2p...).

Reply:
Hrm... there *was* a reason at some point. But i cant recall, and i'm not sure if its still valid. @Kubuxu do you remember?

Reply:
@Stebalien: `secSession.ReadWriter()` forces the handshake, but it doesn't return the error, which we'd definitely want for logging purposes. As a workaround, we could add an error method to the secio `closedRW` and retrieve the error message by type asserting. We wouldn't have to send anything on the wire then, so we could avoid the protocol version bump.

I'd be happy to refactor secio such that the handshake is performed immediately, if this feature isn't needed anymore. I think this would be the cleanest solution here (a handshake takes a couple of RTTs, so latency-wise we're better off doing it asap).

Reply:
I just created a PR in the secio repo to remove the lazy handshake: https://github.com/libp2p/go-libp2p-secio/pull/16.

Reply:
We should be able to remove this now. This PR should work with or without the lazy handshake, and with or without this hacky forced handshake thing.

Reply:
Also, for reference, you can just call `.RemotePeer()` to force the handshake.

Reply:
This shouldn't be needed anymore at all. I just forgot to remove this. PR is updated now.